### DISCUSSION TITLE
Is it possible to export over ssh?

### DISCUSSION THREAD

**Discussion by Juggler00** [RhetTbull/osxphotos#1711] [Q&A] [answered]
Reading the documentation, there are clear examples of exporting to an externally-mounted disk or NAS. Is it possible to specify a ssh destination similar to how rsync works?

e.g. > osxphotos export user@192.168.0.3:/home/photos_backup --retry 3

**Comment by RhetTbull** [comment, marked as answer]
It is not currently possible to do this directly but there is a work-around (see below). Implementing something like this would take some re-architecture of the current code which requires inspection of the exported files to determine if one needs to be updated. OSXPhotos creates an export database (usually in the export folder) that could be stored locally and queried to see which files need to be updated so it's feasible but would require some code changes.  I'm willing to open an issue for this (it would enable other backends like S3) but I'm very busy at the moment so this is likely very far down the road unless someone wants to submit a PR. 

As a work around you could do an export to local disk and use the `--post-command` to export any newly exported or updated photo to the remote server using `rsync`. If you are using "Download originals to this Mac" in Photos and APFS on your Mac filesystem (the default for all Macs for sometime) and you are exporting the unmodified files from Photos (that is, you are not using the `--exiftool` or `--convert-to-jpeg` options) then OSXPhotos will use Copy-on-Write to copy the files out of the Photos library to the target directory. That means the exported files do not take up any additional space on the disk.  I use this regularly to backup my own library so that the exported files can be browsed in Finder and will be backed up independently by Time Machine and Backblaze. Thus there's effectively no "cost" to doing this.

`--post-command exported "rsync -avz --relative {shell_quote,{export_dir}/./{filepath|chomp(35)}} user@192.168.0.3:/home/photos_backup"`

In this example, "exported" causes the command to be applied to any file that has been exported or updated. The OSXPhotos template system is used to specify the file path for rsync and the rsync `--relative` option is used to exclude the local export directory from the remote path. All OSXPhotos templates used in the command will be interpolated by OSXPhotos before being passed to the shell for execution. Here's a breakdown of the template `{shell_quote,{export_dir}/./{filepath|chomp(35)}} `:

`{shell_quote,...}` : use shell quoting for the following value so that spaces in directory or file names are properly quoted or escaped.

`{export_dir}` : the export directory passed to `osxphotos export`; this is the root path we'll want to exclude from the remote path.

`/./` : used to indicate to `rsync --relative` which part of the path should be excluded

`{filepath|chomp(35)}` : `{filepath}` specifies the full path to the exported file. In this case, we want to chop off the beginning part so that only the part relative to the export directory is used. the `|chomp(X)` filter chops off `X` characters from the beginning of the passed value. You will need to specify this for your particular local export directory. Include one extra for the trailing slash. For example, if your export directory is `Users/user/Pictures/photos_backup` you would use `|chomp(35)` to remove the first 35 characters (34 for `/Users/user/Pictures/photos_backup` plus 1 for trailing `/`).

Commands run synchronously so the `rsync` will block further export until it completes.  See also `--post-command-error` which allows  you to control the behavior if there's an rsync error:

```
--post-command-error ACTION     Specify either `continue` or `break` for ACTION to control behavior when a
                                post-command fails. If `continue`, osxphotos will log the error and continue
                                processing. If `break`, osxphotos will stop processing any additional
                                --post-command commands for the current photo but will continue with the
                                export. Without --post-command-error, osxphotos will abort the export if a
                                post-command encounters an error.
```

I tested this on my system and it appeared to work as expected. I am not a big rsync user so I don't know what happens if the target file already exists (for example, the photo was edited in Photos so it is being re-exported). You may need to account for that but I'm sure there's an rsync option.

**Reply by Juggler00** [reply]
Thank you. I am looking to do a backup as a "3 is 2, 2 is 1, 1 is none" scenario. I download all originals to Mac, and forgot that APFS does copy-on-write. This is a perfect implementation!

**Reply by gabeklavans** [reply]
From my understanding of `rsync -a`, it already checks for modified files and you can also have it delete files that are no longer present in the source. I'd guess that you can just run one big `rsync` after a successful osxphotos export to simplify the process. In fact I might try that method, since my current method of exporting to a mounted SMB share just periodically causes me headaches.
